numba/jitclass/base.py

```diff
@@ -0,0 +1,359 @@
+# Registration utils
+
+def register_class_type(cls, spec, class_ctor, builder):
+    """
+    Internal function to create a jitclass.
+
+    Args
+    ----
+    cls: the original class object (used as the prototype)
+    spec: the structural specification contains the field types.
+    class_ctor: the numba type to represent the jitclass
+    builder: the internal jitclass builder
+    """
+    # Normalize spec
+    if isinstance(spec, Sequence):
+        spec = OrderedDict(spec)
+
+    # Copy methods from base classes
+    clsdct = {}
+    for basecls in reversed(inspect.getmro(cls)):
+        clsdct.update(basecls.__dict__)
+
+    methods = dict((k, v) for k, v in clsdct.items()
+                   if isinstance(v, pytypes.FunctionType))
+    others = dict((k, v) for k, v in clsdct.items() if k not in methods)
+    docstring = others.pop('__doc__', "")
+    _drop_ignored_attrs(others)
+    if others:
+        msg = "class members are not yet supported: {0}"
+        members = ', '.join(others.keys())
+        raise TypeError(msg.format(members))
+
+    class_type = class_ctor(cls, spec, methods)
+    cls = JitClassType(cls.__name__, (cls,), dict(class_type=class_type,
+                                                  __doc__=docstring))
+
+    jitmethods = {}
+    for k, v in methods.items():
+        jitmethods[k] = njit(v)
+
+    # Register resolution of the class object
+    typer = CPUTarget.typing_context
+    typer.insert_global(cls, class_type)
+
+    # Register class
+    backend = CPUTarget.target_context
+    builder(class_type, jitmethods, methods, typer, backend).register()
+
+    return cls
+
+
+def _drop_ignored_attrs(dct):
+    # ignore anything defined by object
+    drop = set(['__weakref__',
+                '__module__',
+                '__dict__'])
+    for k, v in dct.items():
+        if isinstance(v, (pytypes.BuiltinFunctionType,
+                          pytypes.BuiltinMethodType)):
+            drop.add(k)
+        elif getattr(v, '__objclass__', None) is object:
+            drop.add(k)
+
+    for k in drop:
+        del dct[k]
+
+class ClassBuilder(object):
+    """
+    A jitclass builder for mutable jitclasses.
+    """
+    instance_type_class = types.ClassInstanceType
+
+    def __init__(self, class_type, jitmethods, methods, typer, backend):
+        self.class_type = class_type
+        self.jitmethods = jitmethods
+        self.methods = methods
+        self.typer = typer
+        self.backend = backend
+
+    def register(self):
+        """
+        Register to the frontend and backend.
+        """
+        outer = self
+
+        class ConstructorTemplate(templates.AbstractTemplate):
+            key = outer.class_type
+
+            def generic(self, args, kws):
+                ctor = outer.jitmethods['__init__']
+
+                instance_type = outer.class_type.instance_type
+                if instance_type not in defined_types:
```

Comment:
No.  It's due to earlier version of allowing some crazy way of specializing the class type.  Since that is removed, we don't need to lazily do this anymore.  Thanks for catching this.